### ISSUE TITLE
Cannot Mock Dependency of Dependency

### ISSUE THREAD

**Issue by travigd** [Urigo/graphql-modules#689]
I'm not sure if there's a different/recommended way to do this, but, I'm unable to mock the dependencies of my dependencies.

Let's say we have providers `A -> B -> C` where `->` means "depends on" (_i.e._ `A` injects `B` into itself and `B` injects `C` into itself). All three are in different modules.

For my test setup, I want to mock `A` and `C`, but not `B` (for my use case, `A` and `C` depend on database access but `B` does not). If I do `injector.provide(...)` for both `A` and `C`, I get an error later when `B` tries to load `C` because it tries to load the non-mocked version of `C`.

The reason for this is that when `A` goes to load `B`, it doesn't find it in the current injector (because it's in a different module) and goes and looks in its children and finds `B` there. From there, `B` uses the child injector to load `C` - but the child injector hasn't had `C` mocked, and thus, here we are.

The not-great solution (imho) is to manually provide `B` (the non-mocked version). It would be nice if the injector resolution wasn't hierarchical like this (instead, a single injector per scope could be used). This would also have the benefit of making DI faster (no need to check all children which is potentially O(number of providers)) though I don't think it would have any real effect on speed.

**Comment by travigd** [comment]
I actually can't even use the workaround I described above wherein you just `injector.provide` everything - it results in application-scope providers being constructed for every session (I can't quite figure out why).

**Comment by travigd** [comment]
Looking into this a bit more (I was frustrated by how long it took me to debug these issues yesterday :')), one possible workaround is just to inject the mock provider using the provider-to-be-mocked's module's injector.

i.e.
```ts
// Same setup as in original post where A injects B which injects C
ModuleA.inject({ provide: ProviderA, useClass: MockProviderA });
ModuleC.inject({ provide: ProviderC, useClass: MockProviderC });
```

The only issue with this is that it's not limited to the test case where I do it. I thought I'd be able to get around it by doing `modA = ModuleA.forChild({})` but it doesn't create new copies of its children.

On a related note, I'm not sure what `forRoot` and `forChild` are supposed to actually do, and the documentation on `forRoot` seems not to be correct.
```ts
  /**
   * Creates another instance of the module using a configuration
   * @param config - the config object
   */
  forRoot(config: Config): GraphQLModule<Config, Session, Context, SelfResolvers> {
    this._moduleConfig = {
      ...this._moduleConfig,
      ...config
    };
    // clean cache
    // ...
    return this;
  }
```

It says that it "creates another instance," but it doesn't (it returns the same instance), whereas the undocumented `createChild` **does** create a new instance (but doesn't create new instances of its children).

All in all, I can get around this (i.e. overwriting providers not being an "isolated" thing) right now by doing the mocks per-file (since Jest does each test file in a new process).

P.S. I haven't had any issues with this project until this and I'm a huge fan of what y'all are doing here! I'd be happy to write a PR that implements the "deep copying" functionality in `forChild` (or a new `deepCopy` method); seems like it would be kind of messy if your dependency graph isn't a tree (i.e. you don't want two separate copies of a module that is depended on by 2+ other modules).

**Comment by ardatan** [comment]
@travigd So in your case `clearCache` might work. Could you try that?

**Comment by travigd** [comment]
Unless I'm misunderstanding, `clearCache` would work to clear the injector cache between test cases in the same file (kind of - I'd still have to clear the cache of each module individually that contains a provider that I mocked) which doesn't really solve the root issue of wanting to mock providers from other modules.

**Comment by kamilkisiela** [comment]
We released a new major version (v1.0) of GraphQL Modules. Here's the [migration guide](https://graphql-modules.com/docs/next/recipes/migration).

The new version solves the issue, so I'm closing the thread.

Please try to upgrade. If you can't and still want to use the legacy version, please open a new issue.
